### PR TITLE
Provide 3rd party license check

### LINKED ISSUE #445: Provide 3rd party license check and sbom
**Describe the feature request**
 Fulfillment of the requirements in the section of 3rd party licenses.

**Upstream changes**
* [ ] sbom is available for every version
* [ ] provide allowlist
* [ ] all licenses are checked against allowlist
* [ ] README for usage of sbom

### PR DESCRIPTION
**Pull request by MartinGroscheTT**
Resolves #445

### CODE REVIEW

File: config/allowed_licenses.json [hunk 1]
```diff
@@ -0,0 +1,2832 @@
...
+        }
+      ]
+    },
+    {
+      "group": "com.jacob",
+      "name": "jacob",
+      "version": "1.20",
+      "licenses": [
+      ]
+    },
```
Comment (cwellmtt) [inline config/allowed_licenses.json:36]: Entry exists twice

File: config/check_dependencies.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        print("{} does not exist.".format(file_path))
+        sys.exit(2)
+
+    with open(file_path) as f:
+        return json.load(f)
+
+
+def validate(allow_list, sbom):
+    allow_json = read_json(allow_list)
+    sbom_json = read_json(sbom)
```
Comment (cwellmtt) [inline config/check_dependencies.py:65]: 2 times validate method - more specific names

File: config/check_dependencies.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    # dependencies
+    if len(allow_json["components"]) != len(sbom_json["components"]):
+        allowed_groups = [item["name"] for item in allow_json["components"]]
+        sbom_groups = [item["name"] for item in sbom_json["components"]]
+        error_collection.append("Number of licenses expects {} but was {}.".format(len(allow_json["components"]),
+                                                                                   len(sbom_json["components"])))
+        missing_allowed = set(sbom_groups).difference(allowed_groups)
+        missing_sbom = set(allowed_groups).difference(sbom_groups)
+        if missing_sbom:
+            error_collection.append("License {} not found in sbom but in allow list.".format(missing_sbom))
```
Comment (cwellmtt) [inline config/check_dependencies.py:83]: If this is an error, is the allow list really an allow list or rather a "checklist"?

File: config/check_dependencies.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    validator = ComponentValidator(allow_json["component"], sbom_json["metadata"]["component"])
+    product_err = validator.validate(True)
+    if product_err:
+        error_collection.append(product_err)
+
+    # dependencies
+    if len(allow_json["components"]) != len(sbom_json["components"]):
+        allowed_groups = [item["name"] for item in allow_json["components"]]
+        sbom_groups = [item["name"] for item in sbom_json["components"]]
+        error_collection.append("Number of licenses expects {} but was {}.".format(len(allow_json["components"]),
```
Comment (cwellmtt) [inline config/check_dependencies.py:78]: actually, number of components

File: config/check_dependencies.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        missing_allowed = set(sbom_groups).difference(allowed_groups)
+        missing_sbom = set(allowed_groups).difference(sbom_groups)
+        if missing_sbom:
+            error_collection.append("License {} not found in sbom but in allow list.".format(missing_sbom))
+        elif missing_allowed:
+            error_collection.append("License {} not found in allow list but in sbom.".format(missing_allowed))
+
+    for component in sbom_json["components"]:
+        if component["name"] not in [component["name"] for component in allow_json["components"]]:
+            error_collection.append("Could not find component '{}' in allow list".format(component["name"]))
```
Comment (cwellmtt) [inline config/check_dependencies.py:89]: Doesn't this check the same as missing_allowed?

File: .reuse/dep5 [hunk 6]
```diff
@@ -2,13 +2,11 @@ Format: https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/
...
 
 Files: *
-Copyright: $YEAR $NAME <$CONTACT>
+Copyright: 2015-2021 TraceTronic GmbH <info@tracetronic.de>
 License: BSD-3-Clause
 
 Files: lib/com/jacob/*
-Copyright: $YEAR $NAME <$CONTACT>
+Copyright: 2020 FreemanSoft Inc <joe@freemansoft.com>
 License: LGPL-2.1-only
```
[the remark this answers is not in the document]
Comment (MartinGroscheTT) [reply .reuse/dep5:12]: Because it did not work really well

File: config/allowed_licenses.json (hunk 1 above)
Comment (MartinGroscheTT) [reply config/allowed_licenses.json:36]: That's right. Otherwise the comparison to the length of license elements cannot be made

**Review by cwellmtt — APPROVED** [review]

**Comment by ErikRehmTT** [comment]
Closing, as original issue was rejected.

**Pull request closed without merging**